### ISSUE TITLE
Run test suites in azure ci

### ISSUE THREAD

**Issue by jneira** [haskell/haskell-ide-engine#1401]
* This is the more promising option for running tests in macos and windows (thanks god and @alanz at least we have the circle ci for linux)
* cc @power-fungus 
* Currently [unit tests are failing](https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=209) due to a `cabal-helper` [issue](https://github.com/DanielG/cabal-helper/issues/78).
  * That behaviour is not present in local or circleci

**Comment by fendor** [comment]
Some of these issues should be fixed by #1126, so maybe we can try again to enable azure-ci testing, once that pr has landed.

**Comment by jneira** [comment]
I am just running the pr in azure with tests enabled: https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=212

**Comment by jneira** [comment]
Hi, after applying the workaround to avoid the errors compiling cabal-helper at runtime the number of failed tests has decreased sensibly:
* [master](https://github.com/jneira/haskell-ide-engine/tree/azure-win-tests): https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=219
   * Quite builds have only errors related with liquid haskell and even a linux build succeded: https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=219&jobId=86dfe75d-acf3-5a7d-391c-10fa0ecb27bb

The functional tests suite with the hie-bios branch have more fails:  https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=221. Most of them are like:

```
haskell-ide-engine>   1) Completion.completions works
haskell-ide-engine>        uncaught exception: ConduitParserException
haskell-ide-engine>        NamedParserException "NotificationMessage ServerMethod PublishDiagnosticsParams" (NamedParserException "NotificationMessage ServerMethod PublishDiagnosticsParams" (NamedParserException "NotificationMessage ServerMethod PublishDiagnosticsParams" (NamedParserException "Logging notification" (BothFailed (Unexpected "ConduitParser.empty") (NamedParserException "NotificationMessage ServerMethod PublishDiagnosticsParams" (NamedParserException "NotificationMessage ServerMethod PublishDiagnosticsParams" (NamedParserException "NotificationMessage ServerMethod PublishDiagnosticsParams" (NamedParserException "Logging notification" (BothFailed (Unexpected "ConduitParser.empty") (NamedParserException "NotificationMessage ServerMethod PublishDiagnosticsParams" (NamedParserException "NotificationMessage ServerMethod PublishDiagnosticsParams" (NamedParserException "NotificationMessage ServerMethod PublishDiagnosticsParams" (NamedParserException "Logging notification" (BothFailed (Unexpected "ConduitParser.empty") (Unexpected "ConduitParser.empty")))))))))))))))
haskell-ide-engine> 
haskell-ide-engine>   To rerun use: --match "/Completion/completions/works/"
```

**Comment by jneira** [comment]
I've rerun tests in azure with actual master: https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=334&view=results
* The unit-test suite passes
* The dispatcher and plugin-dispatcher suites pass
* The functional-test fails due to different behaviour of diagnostics received
* The wrapper-test fails due to different output of the stack command

**Comment by fendor** [comment]
This would be great! maybe we should start poking members of the haskell organizations to enable azure for this project?

**Comment by jneira** [comment]
Absolutly, do we know who has the permissions to do it, @alanz, @bubba?

**Comment by alanz** [comment]
I think the haskell committee has control of this. Not sure who specifically.

**Comment by jneira** [comment]
Ok, maybe we can wait to have the azure setup complete, including green test suites and deploy

**Comment by jneira** [comment]
Update:
* [X] Linux jobs: all test suites are green 
* [ ] Windows jobs
  * [X] unit-test suite: green 
  * [ ] dispatcher-test: flaky, it fails sometimes for all ghcs always with same errors (see below)
  * [X] plugin-dispatcher: green
  * [ ] wrapper-test: failing
  * [ ] functional-test: one failing test
* [ ] macos jobs
  * [X] ghc-8.6.*: all test suites green
  * [ ] ghc-8.4.*: test hangs after error in dispatcher test (see below)

* Executions:
  * https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=344
  * https://dev.azure.com/jneira/haskell-ide-engine/_build/results?buildId=411

-----------

Windows dispatcher-test errors:
```
Failures:

  test\dispatcher\Main.hs:157:5: 
  1) functional dispatch defers responses until module is loaded
       uncaught exception: ErrorCall
       Prelude.head: empty list

  To rerun use: --match "/functional dispatch/defers responses until module is loaded/"

  test\dispatcher\Main.hs:205:7: 
  2) functional dispatch instantly responds to deferred requests if cache is available
       expected: ("references",Just [DocumentHighlight {_range = Range {_start = Position {_line = 7, _character = 0}, _end = Position {_line = 7, _character = 2}}, _kind = Just HkWrite},DocumentHighlight {_range = Range {_start = Position {_line = 7, _character = 0}, _end = Position {_line = 7, _character = 2}}, _kind = Just HkWrite},DocumentHighlight {_range = Range {_start = Position {_line = 5, _character = 6}, _end = Position {_line = 5, _character = 8}}, _kind = Just HkRead},DocumentHighlight {_range = Range {_start = Position {_line = 7, _character = 0}, _end = Position {_line = 7, _character = 2}}, _kind = Just HkWrite},DocumentHighlight {_range = Range {_start = Position {_line = 7, _character = 0}, _end = Position {_line = 7, _character = 2}}, _kind = Just HkWrite},DocumentHighlight {_range = Range {_start = Position {_line = 5, _character = 6}, _end = Position {_line = 5, _character = 8}}, _kind = Just HkRead}])
        but got: ("req2",Nothing)

  To rerun use: --match "/functional dispatch/instantly responds to deferred requests if cache is available/"

  test\dispatcher\Main.hs:257:7: 
  3) functional dispatch returns hints as diagnostics
       expected: ("r5",Just (PublishDiagnosticsParams {_uri = Uri {getUri = "file:///D:/a/1/s/test/testdata/FuncTest.hs"}, _diagnostics = List [Diagnostic {_range = Range {_start = Position {_line = 9, _character = 6}, _end = Position {_line = 10, _character = 18}}, _severity = Just DsInfo, _code = Just (StringValue "Redundant do"), _source = Just "hlint", _message = "Redundant do\nFound:\n  do putStrLn \"hello\"\nWhy not:\n  putStrLn \"hello\"\n", _relatedInformation = Nothing}]}))
        but got: ("references",Nothing)

  To rerun use: --match "/functional dispatch/returns hints as diagnostics/"

  test\dispatcher\Main.hs:297:7: 
  4) functional dispatch instantly responds to failed modules with no cache with the default
       expected: ("req7",Just [])
        but got: ("r5",Nothing)

  To rerun use: --match "/functional dispatch/instantly responds to failed modules with no cache with the default/"
```

-----------

macos error in dispatcher-test for ghc-8.4

```
haskell-ide-engine> dispatcher-test: 
haskell-ide-engine> lookupSymbol failed in relocateSection (RELOC_GOT)
haskell-ide-engine> /Users/runner/runners/2.163.1/work/1/s/.stack/programs/x86_64-osx/ghc-8.4.3/lib/ghc-8.4.3/integer-gmp-1.0.2.0/HSinteger-gmp-1.0.2.0.o: unknown symbol `___gmp_rands'
```

**Comment by jneira** [comment]
* Intermitent errors:
  * stack windows jobs fail with `C:\path\to\ghc\strip.exe: unable to rename 'D:\sr\snapshots\add677e6\bin\happy.exe'; reason: File exists`. I've hit it in local, and you have to delete the snapshot
  * stack windows job fail installing the specific ghc for liquidhaskell: error decompressing tar

**Comment by jneira** [comment]
After #1601 there are a new failing unit test in windows:
* Windows-Stack-8.6.5
```
Failures:

  test/unit\\ApplyRefactPluginSpec.hs:42:5: 
  1) ApplyRefactPlugin, apply-refact plugin, apply-refact plugin commands, applies one hint only
       uncaught exception: GhcException
       unit-test.exe: could not detect mingw toolchain

  To rerun use: --match "/ApplyRefactPlugin/apply-refact plugin/apply-refact plugin commands/applies one hint only/"
```
* Windows-Cabal-8.4.4
```
Failures:

  test/unit\\ApplyRefactPluginSpec.hs:134:5: 
  1) ApplyRefactPlugin, apply-refact plugin, apply-refact plugin commands, respects hlint config files in project root dir
       uncaught exception: IOException of type NoSuchThing
       .\settings: openFile: does not exist (No such file or directory)

  To rerun use: --match "/ApplyRefactPlugin/apply-refact plugin/apply-refact plugin commands/respects hlint config files in project root dir/"
```
